File: hexameter-core/src/main/java/org/codetome/hexameter/core/api/Hexagon.java
package org.codetome.hexameter.core.api;

import java.awt.Rectangle;
import java.util.Collection;
import java.util.List;

import org.codetome.hexameter.core.api.contract.SatelliteData;
import org.codetome.hexameter.core.backport.Optional;

/**
 * Represents a Hexagon.
 * <em>Please note</em> that all coordinates are relative to the {@link HexagonalGrid} containing this {@link Hexagon}.
 */
public interface Hexagon<T extends SatelliteData> {

    /**
     * Returns an unique {@link String} representing this {@link Hexagon}.
     */
    String getId();

    /**
     * Returns a list containing the {@link Point}s of this {@link Hexagon}.
     */
    Collection<Point> getPoints();
    
    /**
     * Returns a list containing the {@link Point}s of this {@link Hexagon}.
     */
    List<Point> getPointList();
    
    /**
     * Returns an array containing the {@link Point}s of this {@link Hexagon}
     */
    Point[] getPointArray();
    
    /**
     * Returns an array of the vertices of this {@link Hexagon}
     */
    float[] getVertices();
    
    /**
     * Returns a rectangle defining the <b>external</b> boundary box of this {@link Hexagon} (a rectangle that hits the 2 pointy corners and the 2 flat sides)
     */
    Rectangle getExternalBoundingBox();
    
    /**
     * Returns a rectangle defining the <b>internal</b> boundary box of this {@link Hexagon} (the biggest rectangle that hits the outline of the {@link Hexagon} exactly 4 times)
     */
    Rectangle getInternalBoundingBox();

    /**
     * Returns the {@link CubeCoordinate} of this {@link Hexagon}.
     */
    CubeCoordinate getCubeCoordinate();

    /**
     * Returns this {@link Hexagon}'s <b>x</b> (cube) coordinate on the {@link HexagonalGrid}.
     */
    int getGridX();

    /**
     * Returns this {@link Hexagon}'s <b>y</b> coordinate on the {@link HexagonalGrid}.
     * The Y coordinate is not present in the cube model but it is in the cube model.
     * This method is just for convenience.
     */
    int getGridY();

    /**
     * Returns this {@link Hexagon}'s <b>z</b> (cube) coordinate on the {@link HexagonalGrid}.
     */
    int getGridZ();

    /**
     * Returns the center <b>x</b> (pixel) coordinate of this {@link Hexagon}.
     */
    double getCenterX();

    /**
     * Returns the center <b>y</b> (pixel) coordinate of this {@link Hexagon}.
     *
     * @return center y
     */
    double getCenterY();

    /**
     * Returns this {@link Hexagon}'s satellite data.
     */
    Optional<T> getSatelliteData();

    /**
     * Can be used to add arbitrary satellite data to a {@link Hexagon}.
     */
    void setSatelliteData(T data);

    /**
     * Clears the satellite data of this Hexagon.
     */
    void clearSatelliteData();

}

File: mixite2.core/src/commonMain/kotlin/org/hexworks/mixite2/core/api/Hexagon.kt
package org.hexworks.mixite2.core.api

import org.hexworks.cobalt.datatypes.Maybe
import org.hexworks.mixite2.core.api.contract.SatelliteData

/**
 * Represents a Hexagon.
 * *Please note* that all coordinates are relative to the [HexagonalGrid] containing this [Hexagon].
 */
interface Hexagon<T : SatelliteData> {

    /**
     * Returns an unique [String] representing this [Hexagon].
     */
    val id: String

    /**
     * Returns a list containing the [Point]s of this [Hexagon].
     */
    val points: List<Point>

    /**
     * Returns an array of the vertices of this [Hexagon].
     */
    val vertices: List<Double>

    /**
     * Returns a rectangle defining the **external** boundary box of this [Hexagon]
     * (a rectangle that hits the 2 pointy corners and the 2 flat sides).
     */
    val externalBoundingBox: Rectangle

    /**
     * Returns a rectangle defining the **internal** boundary box of this [Hexagon]
     * (the biggest rectangle that hits the outline of the [Hexagon] exactly 4 times).
     */
    val internalBoundingBox: Rectangle

    /**
     * Returns the [CubeCoordinate] of this [Hexagon].
     */
    val cubeCoordinate: org.hexworks.mixite2.core.api.CubeCoordinate

    /**
     * Returns this [Hexagon]'s **x** (cube) coordinate on the [HexagonalGrid].
     */
    val gridX: Int

    /**
     * Returns this [Hexagon]'s **y** coordinate on the [HexagonalGrid].
     * The Y coordinate is not present in the cube model but it is in the cube model.
     * This method is just for convenience.
     */
    val gridY: Int

    /**
     * Returns this [Hexagon]'s **z** (cube) coordinate on the [HexagonalGrid].
     */
    val gridZ: Int

    /**
     * Returns the center (pixel) [Point] of this [Hexagon].
     */
    val center: Point

    /**
     * Returns this [Hexagon]'s satellite data.
     */
    val satelliteData: Maybe<T>

    /**
     * Can be used to add arbitrary satellite data to a [Hexagon].
     */
    fun setSatelliteData(data: T)

    /**
     * Clears the satellite data of this Hexagon.
     */
    fun clearSatelliteData()

}
